# 4.7 IC Deriving new variables

# Table of Contents

#### 01. Import libraries

#### 02. Deriving new variables

#### 02.01 Define a function

#### 02.01(a) If-elif-else

#### 02.01(b) Use .loc() function

#### 02.02 Use a for-loop to create a new column

#### 03. Task 4.7 Deriving new variables

#### Step 2 - Add a new column using a for-loop

#### Step 4 - Add a new column using a .loc() function

#### Step 7 - Export data as a pickle file

# 01. Import libraries

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Deriving new variables

In [7]:
# Import ords_prods_merge df from pickle file previously saved

path = r'C:\Users\dirk8\CareerFoundry Projects\03-2025 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [9]:
# Check the output
ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0


In [11]:
# Check the dimensions
ords_prods_merge.shape

(32404859, 13)

In [13]:
# Create a subset of the ords_prods_merge df containing the first 1 million records, for Exercise 4.7

df = ords_prods_merge[:1000000]

In [15]:
# Check created subset df
df.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0


In [17]:
# Check the dimensions
df.shape

(1000000, 13)

# 02.01 Define a function

# 02.01(a) If-elif-else

In [23]:
# Define a function

def price_label(row):
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return np.nan

In [27]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\dirk8\AppData\Local\Temp\ipykernel_22424\93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [29]:
# Run .value_counts() function to check values frequency, use dropna = False to included any NaN value counts

df['price_range'].value_counts(dropna = False)

price_range
Mid range product     673183
Low range product     314392
High range product     12425
Name: count, dtype: int64

In [31]:
# Check subset dataframe
df.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid range product
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid range product
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low range product
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low range product
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low range product


In [33]:
# Check the max value in the subset dataframe
df['prices'].max()

99999.0

**Comment: The 'prices' value = 99999 must be an outlier, possibly to be removed in a later Exercise or Task after visualization or any further descriptive stats analysis.  If it is not flagged in a later Exercise, then I will circle back to remove it and any other values that are more than 2 standard deviations away from the 'prices' mean value.**

# 02.01(b) Use .loc() function

In [43]:
# As per the Error message above, use Python's predefined loc() function to accomplish the same defined function nested if-then result
# Note that the lines substituting for if-then execution should be in separate cells

In [49]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [50]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [51]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [52]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

In [81]:
# Note to self: when I ran the first line cell above it gave an error message, but the 3 lines together did work for some reason
# Also, not clear why we do not need a 4th line that mimics the "else: return np.nan" in our equivalent custom-defined function example

In [83]:
# Check the output
df.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range,price_range_loc
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid range product,Mid-range product
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid range product,Mid-range product
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low range product,Low-range product
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low range product,Low-range product
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low range product,Low-range product


# Use loc() for same result as custom-defined function but saves RAM

In [88]:
# Repeat the loc() use example above but on the entire ords_prods_merge dataframe as it is RAM-friendly
# Check the frequency distribution of the created price_range_loc column values

In [89]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [90]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <=15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [91]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [92]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [162]:
# Check the output

ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product


In [164]:
# Check the dimensions
ords_prods_merge.shape

(32404859, 14)

In [166]:
# We will be using for-loop to create a new column summarizing how busy each day of the week is
# First, we need to know on which day most orders take place, e.g. a frequency distribution

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [170]:
# Based on the above frequencies (transaction counts), Saturday (= 0) is the busiest day and Wednesday (= 4) the least busiest day

# 02.02 Use a for-loop to create a new column

# Use for-loop to create a new column classifying how busy each day of the week is (based on transaction counts)

In [273]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")                    

In [274]:
# Check first 5 elements of the created "result" list
result[:5]

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy']

NOTE: The for-loop syntax followed the Exercise's example of using " (double quotes) rather than ' (single quotes) but the for-loop operation still worked

In [277]:
# Need to add the "result" list as a new column in the dataframe

ords_prods_merge['busiest_day'] = result

In [279]:
# Check the frequency of the new column's values and compare with the orders_day_of_week frequencies output earlier (above)

ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [281]:
# Check altered dataframe output
ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_days
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Regularly busy
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Regularly busy
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Regularly busy
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Regularly busy
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Regularly busy


In [283]:
# Check the altered dataframe dimensions
ords_prods_merge.shape

(32404859, 16)

# 03. Task 4.7 Deriving new variables

# Step 2 - Add a new column using a for-loop

In [ ]:
# Add a new column that relabels transaction frequency counts to expand inclusion to the 2 busiest days and 2 slowest days
# From prior results, the 2 busiest days are Saturday (=0) and Sunday (=1), two slowest day are Tuesday (=3) and Wednesday (=4)

In [192]:
outcome = []

for x in ords_prods_merge["orders_day_of_week"]:
    if x in [0, 1]:
        outcome.append("Busiest days")
    elif x in [3, 4]:
        outcome.append("Slowest days")
    else:
        outcome.append("Regularly busy")   

In [193]:
# Check first 5 elements of the created "outcome" list
outcome[:5]

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy']

In [254]:
# Need to add the "outcome" list as a new column labeled 'busiest_days' (plural form) in the dataframe

ords_prods_merge['busiest_days'] = outcome

# Step 3

In [261]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [263]:
# Check the frequency of the new column's values and compare with the orders_day_of_week frequencies output above

ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

# The relabeled transaction counts match exactly the sum of the relevant day of week transaction counts in the above frequency distribution for each day of the week

In [248]:
# Check the ords_prods_merge output
ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_days
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Regularly busy
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Regularly busy
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Regularly busy
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Regularly busy
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Regularly busy


In [252]:
# Check the dimensions
ords_prods_merge.shape

(32404859, 16)

# Comment: In Step 2, if my RAM were more limited, I would have used an loc() function-based variable derivation rather than the for-loop approach

# Step 4 - Add a new column using .loc() function

In [268]:
# Step 4: Create a new column identifying busiest hours of day broken down into 3 periods, "Most orders", "Average orders" and "Fewest orders"
# We start by looking at the frequency distribution of the 'order_hour_of_day' column to figure out period labeling

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

__Based on the above frequency distribution we label time periods as follows: "Most orders" is order_hour_of_day whose transaction count > 2000000, "Average orders" is < 2000000 and >= 1000000, and "Fewest orders" is < 1000000.  We use an loc() function-based variable derivation rather than a for-loop approach to save RAM usage. Then we output a frequency distribution.__

In [293]:
ords_prods_merge.loc[ords_prods_merge['order_hour_of_day'].isin([9,10,11,12,13,14,15,16,17]), 'busiest_period_of_day'] = 'Most orders'

In [294]:
ords_prods_merge.loc[ords_prods_merge['order_hour_of_day'].isin([8,18,19]), 'busiest_period_of_day'] = 'Average orders'

In [295]:
ords_prods_merge.loc[ords_prods_merge['order_hour_of_day'].isin([0,1,2,3,4,5,6,7,20,21,22,23]), 'busiest_period_of_day'] = 'Fewest orders'

# Step 5

In [342]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       23205725
Average orders     4612925
Fewest orders      4586209
Name: count, dtype: int64

In [385]:
# Check the altered dataframe

ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_days,busiest_period_of_day
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Regularly busy,Average orders
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Regularly busy,Average orders
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Regularly busy,Average orders
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Regularly busy,Average orders
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Regularly busy,Average orders


In [387]:
# Check the dimensions
ords_prods_merge.shape

(32404859, 17)

In [389]:
# Check summary stats

ords_prods_merge.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,aisle_id,department_id,prices
count,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.032876e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07
mean,1.710745e+06,1.029372e+05,1.714230e+01,2.738867e+00,1.342515e+01,1.110408e+01,2.559866e+04,8.352547e+00,5.895873e-01,7.119612e+01,9.919792e+00,1.198023e+01
std,9.872988e+05,5.946610e+04,1.753532e+01,2.090077e+00,4.246380e+00,8.779064e+00,1.408400e+04,7.127071e+00,4.919087e-01,3.821139e+01,6.281485e+00,4.956554e+02
min,2.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00
25%,8.559470e+05,5.142200e+04,5.000000e+00,1.000000e+00,1.000000e+01,5.000000e+00,1.354400e+04,3.000000e+00,0.000000e+00,3.100000e+01,4.000000e+00,4.200000e+00
50%,1.711049e+06,1.026160e+05,1.100000e+01,3.000000e+00,1.300000e+01,8.000000e+00,2.530200e+04,6.000000e+00,1.000000e+00,8.300000e+01,9.000000e+00,7.400000e+00
75%,2.565499e+06,1.543890e+05,2.400000e+01,5.000000e+00,1.600000e+01,1.500000e+01,3.794700e+04,1.100000e+01,1.000000e+00,1.070000e+02,1.600000e+01,1.130000e+01
max,3.421083e+06,2.062090e+05,9.900000e+01,6.000000e+00,2.300000e+01,3.000000e+01,4.968800e+04,1.450000e+02,1.000000e+00,1.340000e+02,2.100000e+01,9.999900e+04


# The 'prices' column has a minimum value of 1 and maximum value of 99999. Mean is 12 (rounded) and standard deviation at 496 (rounded) is far too large, which means we have large price outlier(s) skewing the metrics. Also, there should be product prices between 0 and 1, but the minimum is 1 which bears investigation.  Will wait to see if later Exercises have us address these issues with 'prices' data.

# Step 7 - Export data as a pickle file

In [394]:
# Step 7: Export your dataframe as a pickle file (since you added new columns) and store in your "Prepared Data' folder

ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_2.pkl'))